include/mc_trajectory/SequenceInterpolator.h

```diff
@@ -0,0 +1,161 @@
+   *
+   * You must call the \ref setValues "values" setter before calling compute()
+   */
+  SequenceInterpolator() noexcept {}
+
+  /**
+   * @brief Creates an interpolator with values
+   *
+   * @param values Values must respect the conditions describes in \ref setValues "values"
+   *
+   * @throws std::runtime_error If values are invalid
+   */
+  SequenceInterpolator(const TimedValueVector & values)
+  {
+    this->values(values);
+  }
+
+  /**
+   * \anchor setValues
+   *
+   * @brief Set interpolator values.
+   *
+   * @param values Pairs of (time, value) ordered by strictly ascending time and
+   * spread apart by more than dt
+   */
+  void values(const TimedValueVector & values)
+  {
+    if(values.empty()) return;
+    if(!std::is_sorted(values.begin(), values.end(),
+                       [](const TimedValue & a, const TimedValue & b) { return a.first < b.first; }))
+    {
+      mc_rtc::log::error_and_throw<std::runtime_error>(
+          "SequenceInterpolator values must be ordered by strictly ascending time");
+    }
+    values_ = values;
+  }
+
+  /** Interpolation values */
+  inline const TimedValueVector & values() const noexcept
+  {
+    return values_;
+  }
+
+  /**
+   * When true, this interpolator is in a valid state and \ref compute will
+   * always return a valid value.
+   */
+  inline bool hasValues() const noexcept
+  {
+    return values_.size();
+  }
+
+  /** Clears all values */
+  void clear()
+  {
+    values_.clear();
+  }
+
+  /**
+   * Compute interpolated value at the provided time
+   *
+   * This function first finds the current interval corresponding to the
+   * provided time (complexity O(N) at worst), then calls
+   * InterpolationFunction::operator(const Value &, const Value &, double ratio)
+   * where ratio is the time in the current interval normalized between 0 and 1
+   *
+   * @param currTime Time at which to compute the interpolation. Out-of-bound access return the first or last value.
+   *
+   * @throws std::runtime_error If values() is empty
+   */
+  Value compute(double currTime)
+  {
+    if(values_.empty())
+    {
+      mc_rtc::log::error_and_throw<std::runtime_error>("SequenceInterpolator requires at least one value");
+    }
+
+    // Check for out-of-bound access
+    if(currTime >= values_.back().first)
+    {
+      return values_.back().second;
+    }
+    else if(currTime < values_.front().first)
+    {
+      return values_.front().second;
+    }
+
+    // find the current interval
+    nextIndex_ = 0;
+    while(++nextIndex_ < values_.size() && values_[nextIndex_].first < currTime)
+    {
+    }
+    auto prevIndex = static_cast<size_t>(nextIndex_ - 1);
+    double intervalDuration = values_[nextIndex_].first - values_[prevIndex].first;
+
+    const auto & prevTime = values_[prevIndex].first;
+    return interpolator_(values_[prevIndex].second, values_[nextIndex_].second,
+                         (currTime - prevTime) / intervalDuration);
+  }
```

Comment:
You can do a binary search here because values_ is sorted

```c++
size_t prevIndex = 0;
size_t nextIndex = values.size();
while(nextIndex - prevIndex != 1)
{
  size_t pivot = (prevIndex + nextIndex) / 2;
  if(values_[pivot].first < currTime)
  {
    prevIndex = pivot;
  }
  else
  {
    nextIndex = pivot;
  }
}
```

It doesn't make a huge difference when the number of values is small (< 10) but it does significantly improve the search when it gets bigger.

Another simple optimization is to cache the result of the previous search because we now the function is likely to be called with "consecutive" values (i.e. t + dt). We can keep `prevIndex` in the class and check:
- if `values_[prevIndex].first < currTime && values_[prevIndex + 1].first >= currTime` then return `prevIndex`*
- otherwise check the next range (i.e. if `values_[prevIndex + 1].first < currTime && values_[prevIndex + 2].first >= currTime` then return `++prevIndex`*
- otherwise lookup the value

Note that this is independent of doing a linear or binary search and I think this will bypass the search most of the time in our use-case.

*: I don't think we need to check the index in either cases, in the first one because by construction `prevIndex + 1 < values_.size()` and in the second case because the only case where `prevIndex + 2 >= values_.size()` is when the requested time is out of the interpolation range and this is covered before

Reply:
Thanks for the excellent suggestion!
I have implemented it using `std::lower_bound` instead as I'd rather avoid re-implementing STL algorithms, and further cached `nextIndex_` and `intervalDuration_` that only need to be computed when the index actually changes.